File: packages/lattice-rule-sdk-go/artifact_test.go
// The committed artifact and its recorded digest cannot drift apart.
//
// examples/forbidden_tag_dependency.wasm is a binary in a repository that
// reviews text, and …​.wasm.sha256 beside it is the string a customRules row
// pins it by — the mechanism that makes "the law CI ran is the law review saw"
// checkable (../../docs/adr/0002-custom-rules-one-contract.md). Two files that
// must agree and no gate comparing them is a pair that silently stops agreeing
// the first time one of them is rebuilt alone, and the failure surfaces in a
// CONSUMER's tree as a load error naming a hash nobody here can explain.
//
// The Rust binding writes SHA-256 out by hand rather than take a third
// dependency for one test, and holds it honest against FIPS 180-4's published
// vectors (../lattice-rule-sdk-rust/tests/artifact.rs). Nothing like that is
// needed here: crypto/sha256 is the standard library, and this module's
// dependency list is empty either way.

package latticerule

import (
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"os"
	"strings"
	"testing"
)

// artifactPath and digestPath are relative to this package's directory, which
// is where `go test` runs a package's tests from — the module root, for the
// root package.
const (
	artifactPath = "examples/forbidden_tag_dependency.wasm"
	digestPath   = "examples/forbidden_tag_dependency.wasm.sha256"
)

func TestTheCommittedArtifactHashesToTheRecordedDigest(t *testing.T) {
	artifact, err := os.ReadFile(artifactPath)
	if err != nil {
		t.Fatalf("%s could not be read: %v", artifactPath, err)
	}
	recorded, err := os.ReadFile(digestPath)
	if err != nil {
		t.Fatalf("%s could not be read: %v", digestPath, err)
	}

	digest := recorded
	trimmed := strings.TrimSpace(string(digest))
	computed := sha256.Sum256(artifact)
	if hex.EncodeToString(computed[:]) != trimmed {
		t.Fatalf("the committed artifact and its recorded digest have drifted apart — rebuild both "+
			"with ./rebuild-example.sh rather than editing either one\n artifact: %s\n recorded: %s",
			hex.EncodeToString(computed[:]), trimmed)
	}
	if len(trimmed) != 64 {
		t.Fatalf("a sha256 is 64 hex characters, and this is %d", len(trimmed))
	}
	for _, character := range trimmed {
		if !strings.ContainsRune("0123456789abcdef", character) {
			t.Fatalf("the recorded digest must be lowercase hex, which is how a policy row spells "+
				"one: %s", trimmed)
		}
	}
}

// TestTheCommittedArtifactIsAWebAssemblyModule is a four-byte check that
// catches the failures a digest cannot describe: a truncated file, a text
// placeholder, a pointer left by a large-file filter. All three would hash
// consistently and load as nothing.
func TestTheCommittedArtifactIsAWebAssemblyModule(t *testing.T) {
	artifact, err := os.ReadFile(artifactPath)
	if err != nil {
		t.Fatalf("%s could not be read: %v", artifactPath, err)
	}
	if len(artifact) < 8 {
		t.Fatalf("%s is %d bytes, which is not a module", artifactPath, len(artifact))
	}
	magic := [8]byte{0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00}
	if [8]byte(artifact[:8]) != magic {
		t.Fatalf("the artifact does not open with the wasm magic and version 1: % x", artifact[:8])
	}
}

// TestTheCommittedArtifactCarriesNoImportSection is the contract's own refusal,
// checked here on the bytes rather than only by the host.
//
// A module that imports anything is refused at load — "the contract grants no
// imports" — and that is the mechanism behind a rule holding no ambient
// capability. Go cannot instantiate the artifact without a wasm runtime this
// module refuses to depend on, so what it can check is the binary: section id 2
// is the import section, and its absence is what the host's check comes down
// to. ./rebuild-example.sh drives the real instantiation before it records a
// digest, and ./README.md records the round trip through the engine's host.
func TestTheCommittedArtifactCarriesNoImportSection(t *testing.T) {
	artifact, err := os.ReadFile(artifactPath)
	if err != nil {
		t.Fatalf("%s could not be read: %v", artifactPath, err)
	}
	if err := checkNoImportSection(artifact); err != nil {
		t.Fatal(err)
	}
}

// checkNoImportSection walks a module's sections and answers why the contract
// would refuse it, or nil when every section clears.
//
// The walk split from the committed file for the same reason the SHA-256 above
// is held against FIPS 180-4's vectors: the committed artifact carries NO import
// section, so a walk tested only against it can drift — an id comparison that
// stops comparing, a length read that miscounts — and stay green forever by not
// looking. The synthetic modules in
// TestTheSectionWalkRefusalsAreProvenAgainstSyntheticModules are what make this
// function's refusals real rather than unreachable branches.
//
// Sections follow the 8-byte preamble, each a one-byte id then a LEB128 length
// then that many bytes. Walking them needs no wasm knowledge beyond that, and
// stopping at the first malformed length keeps a corrupt file from being read
// as one that simply has no imports; the tail check keeps one whose declared
// lengths run past its end from passing either.
func checkNoImportSection(module []byte) error {
	offset := 8
	for offset < len(module) {
		id := module[offset]
		offset++
		size, read := uvarint(module[offset:])
		if read <= 0 {
			return fmt.Errorf("the section at byte %d has no readable length, so this module "+
				"is not one this test can clear", offset)
		}
		if id == 2 {
			return errors.New("the module declares an import section, and the contract grants " +
				"no imports — the host would refuse it at load")
		}
		offset += read + int(size)
	}
	if offset != len(module) {
		return fmt.Errorf("the sections do not add up to the module's %d bytes, ending at %d",
			len(module), offset)
	}
	return nil
}

// sectionWalkPreamble is the eight bytes every core WebAssembly module starts
// with, the start of every synthetic module below.
var sectionWalkPreamble = []byte{0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00}

// TestTheSectionWalkRefusalsAreProvenAgainstSyntheticModules drives
// checkNoImportSection over modules built here, because the committed artifact
// can only ever show the clearing half of the walk. A refusal that stopped
// firing — on an id compare that regressed, a uvarint that miscounted — would
// leave that artifact green while every module the host actually refuses
// sailed through, which is the silent direction this suite exists to refuse.
func TestTheSectionWalkRefusalsAreProvenAgainstSyntheticModules(t *testing.T) {
	cases := []struct {
		name    string
		module  []byte
		refusal string
	}{
		{
			name: "an import section is refused",
			// One memory import on behalf of the (empty-named) module "a":
			// count=1, name len=0, kind=2, limits flags/min/max.
			module:  append(append([]byte{}, sectionWalkPreamble...), 2, 6, 1, 0, 2, 1, 1, 1),
			refusal: "declares an import section",
		},
		{
			name: "a length that never terminates is refused, not cleared",
			// A continuation byte with nothing after it: reading it as zero
			// sections would be a clean answer over a corrupt file.
			module:  append(append([]byte{}, sectionWalkPreamble...), 1, 0x80),
			refusal: "no readable length",
		},
		{
			name: "a payload truncated after its declared length is refused",
			// Ten of the two hundred bytes the length claims: the framing
			// arithmetic must notice, not wrap around into a clean pass.
			module: append(append(append([]byte{}, sectionWalkPreamble...), 1, 0xc8, 0x01),
				make([]byte, 10)...),
			refusal: "do not add up",
		},
	}
	for _, testCase := range cases {
		t.Run(testCase.name, func(t *testing.T) {
			err := checkNoImportSection(testCase.module)
			if err == nil {
				t.Fatal("the walk cleared a module it must refuse")
			}
			if !strings.Contains(err.Error(), testCase.refusal) {
				t.Fatalf("refused, but not for the reason under test: %v", err)
			}
		})
	}
}

// TestTheSectionWalkClearsAModuleWithoutAnImportSection is the positive half:
// the walk must clear what should clear, including a section whose length
// needs two LEB128 bytes — a uvarint that stopped after the first byte would
// misread it as a tiny section and either fail the framing or skip wrong, and
// only a module that exercises both bytes can tell.
func TestTheSectionWalkClearsAModuleWithoutAnImportSection(t *testing.T) {
	module := append([]byte{}, sectionWalkPreamble...)
	module = append(module, 0, 4, 1, 'a', 'b', 'c') // a custom section
	long := append([]byte{1, 0xc8, 0x01}, make([]byte, 200)...)
	module = append(module, long...) // type section, 200-byte payload, two LEB128 bytes

	if err := checkNoImportSection(module); err != nil {
		t.Fatalf("the walk refused a module with no import section: %v", err)
	}
}

// uvarint reads one LEB128-encoded unsigned integer, answering its value and
// how many bytes it took. read is 0 when the encoding runs off the end.
//
// encoding/binary's Uvarint is the same algorithm and would do; it is written
// out because the answer this test turns on is "how many bytes", and
// binary.Uvarint returns a negative count for two different failures that both
// have to be treated as one here anyway.
func uvarint(data []byte) (value uint64, read int) {
	var shift uint
	for index, b := range data {
		value |= uint64(b&0x7f) << shift
		if b&0x80 == 0 {
			return value, index + 1
		}
		shift += 7
		if shift >= 64 {
			return 0, 0
		}
	}
	return 0, 0
}

File: packages/archkeep-rule-sdk-ts/test/artifact.test.mjs
/**
 * The committed artifact: its digest, its shape, and the exact bytes it answers
 * with across the ABI.
 *
 * `../examples/forbidden-tag-dependency.wasm` is a binary in a repository that
 * reviews text, and `….wasm.sha256` beside it is the string a `customRules` row
 * pins it by — the mechanism that makes "the law CI ran is the law review saw"
 * checkable (`../../../docs/adr/0002-custom-rules-one-contract.md`). Two files
 * that must agree and no gate comparing them is a pair that silently stops
 * agreeing the first time one of them is rebuilt alone, and the failure surfaces
 * in a CONSUMER's tree as a load error naming a hash nobody here can explain.
 *
 * This file drives the module DIRECTLY — `new WebAssembly.Instance`, write the
 * bundle, unpack the `i64` — where `./golden.test.mjs` drives it through the
 * engine's real host. The split is deliberate and both halves are needed: the
 * host parses the verdict before handing it over, so the only place the BYTES
 * that cross the ABI can be pinned is here, and the only place the host's
 * refusals can be exercised is there. A change that reordered or renamed a
 * verdict key would still be a change to what the host reads, and this file is
 * what notices it.
 */

import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import test from "node:test";
import assert from "node:assert/strict";

/** @param {string} relative @returns {string} */
const packagePath = (relative) => fileURLToPath(new URL(`../${relative}`, import.meta.url));

const ARTIFACT = packagePath("examples/forbidden-tag-dependency.wasm");
const DIGEST_FILE = packagePath("examples/forbidden-tag-dependency.wasm.sha256");

/** The four symbols the ABI names, and the kind each must be. */
const REQUIRED_EXPORTS = {
  memory: "memory",
  archkeep_alloc: "function",
  archkeep_describe: "function",
  archkeep_evaluate: "function",
};

test("the committed artifact hashes to the recorded digest", () => {
  const bytes = readFileSync(ARTIFACT);
  const recorded = readFileSync(DIGEST_FILE, "utf8").trim();

  assert.equal(
    createHash("sha256").update(bytes).digest("hex"),
    recorded,
    "the committed artifact and its recorded digest have drifted apart — rebuild both with " +
      "./rebuild-example.sh rather than editing either one",
  );
  assert.equal(recorded.length, 64, "a sha256 is 64 hex characters");
  assert.match(
    recorded,
    /^[0-9a-f]{64}$/,
    "the recorded digest must be lowercase hex, which is how a policy row spells one",
  );
});

test("the committed artifact opens with the wasm magic", () => {
  // A four-byte check that catches the failures a digest cannot describe: a
  // truncated file, a text placeholder, a pointer left by a large-file filter.
  // All three would hash consistently and load as nothing.
  const bytes = readFileSync(ARTIFACT);
  assert.deepEqual(
    [...bytes.subarray(0, 8)],
    [0x00, 0x61, 0x73, 0x6d, 0x01, 0x00, 0x00, 0x00],
    "the artifact does not open with the wasm magic and version 1",
  );
});

test("the committed artifact declares no imports at all", () => {
  // The mechanism behind "a rule holds no ambient capability", checked on the
  // bytes rather than on the build flags that produced them: an AssemblyScript
  // module imports `env.abort` unless it is built with `--use abort=`
  // (`../rebuild-example.sh` argues the flag), and the host refuses any module
  // whose import section is non-empty. A rebuild that dropped the flag would
  // produce an artifact that still passes every other test in this file.
  const module = new WebAssembly.Module(readFileSync(ARTIFACT));
  assert.deepEqual(
    WebAssembly.Module.imports(module),
    [],
    "the artifact declares an import, and the contract grants none",
  );
});

test("the committed artifact exports the four symbols the ABI names", () => {
  const module = new WebAssembly.Module(readFileSync(ARTIFACT));
  const declared = new Map(
    WebAssembly.Module.exports(module).map((entry) => [entry.name, entry.kind]),
  );
  for (const [name, kind] of Object.entries(REQUIRED_EXPORTS)) {
    assert.equal(declared.get(name), kind, `the artifact exports no ${kind} named "${name}"`);
  }
});

/**
 * One call into the module, driven the way the host drives it.
 *
 * @param {WebAssembly.Instance} instance
 * @param {bigint} packed The `i64` an ABI export answered with.
 * @returns {string} The UTF-8 text at the range it named.
 */
function readPacked(instance, packed) {
  assert.equal(typeof packed, "bigint", "an ABI export answers with the packed i64");
  const unsigned = BigInt.asUintN(64, packed);
  const pointer = Number(unsigned >> 32n);
  const length = Number(unsigned & 0xffffffffn);
  const memory = /** @type {WebAssembly.Memory} */ (instance.exports.memory);
  assert.ok(
    pointer + length <= memory.buffer.byteLength,
    `the range ${pointer}..${pointer + length} leaves the module's own linear memory`,
  );
  return Buffer.from(new Uint8Array(memory.buffer).slice(pointer, pointer + length)).toString(
    "utf8",
  );
}

/**
 * The module's answer over one evidence document, as the raw bytes it returned.
 *
 * @param {string} fixture A file under `../fixtures/`.
 * @returns {string}
 */
function evaluateRaw(fixture) {
  const module = new WebAssembly.Module(readFileSync(ARTIFACT));
  const instance = new WebAssembly.Instance(module, {});
  const evidence = readFileSync(packagePath(`fixtures/${fixture}`));
  const alloc = /** @type {(len: number) => number} */ (instance.exports.archkeep_alloc);
  const pointer = alloc(evidence.byteLength);
  new Uint8Array(/** @type {WebAssembly.Memory} */ (instance.exports.memory).buffer).set(
    evidence,
    pointer,
  );
  const evaluate = /** @type {(ptr: number, len: number) => bigint} */ (
    instance.exports.archkeep_evaluate
  );
  return readPacked(instance, evaluate(pointer, evidence.byteLength));
}

test("the describe bytes are the ones that cross the ABI", () => {
  const module = new WebAssembly.Module(readFileSync(ARTIFACT));
  const instance = new WebAssembly.Instance(module, {});
  const describe = /** @type {() => bigint} */ (instance.exports.archkeep_describe);
  assert.equal(
    readPacked(instance, describe()),
    '{"contract":1,"name":"forbidden-tag-dependency","needs":["model","graph"],"findings":' +
      '[{"id":"dependency-on-forbidden-tag","message":"a project depends on a project carrying ' +
      'the forbidden tag"}]}',
  );
});

test("the verdict bytes are the ones that cross the ABI", () => {
  // The one place the exact BYTES are pinned rather than the document.
  // `./golden.test.mjs` compares parsed verdicts, because key order is a
  // serializer's business and a conformance suite four SDKs must reproduce
  // cannot turn on one of them. This test exists anyway: the bytes are what
  // crosses the ABI, and it is byte-for-byte the string the Rust SDK's own
  // suite pins over the same fixture
  // (`../../archkeep-rule-sdk-rust/tests/golden.rs`).
  assert.equal(
    evaluateRaw("edge-into-forbidden-tag.json"),
    '{"contract":1,"verdict":"fail","findings":[{"id":"dependency-on-forbidden-tag","message":' +
      '"beta depends on gamma, which carries \\"layer-infrastructure\\"","sourceFile":' +
      '"packages/beta/src/store.rs","project":"beta"}]}',
  );
});

test("a bundle the rule cannot read becomes unknown rather than pass", () => {
  // The silent direction, at the ABI. Driven here rather than through the host
  // because the host writes a bundle IT built, and what has to be proven is
  // what the module does with bytes that are not one.
  const module = new WebAssembly.Module(readFileSync(ARTIFACT));
  for (const bytes of [
    Buffer.from("not json at all", "utf8"),
    Buffer.from('{"contract":1,"rule":', "utf8"),
    Buffer.from('{"contract":7,"rule":{"name":"x","params":{}}}', "utf8"),
    Buffer.from("", "utf8"),
  ]) {
    const instance = new WebAssembly.Instance(module, {});
    const alloc = /** @type {(len: number) => number} */ (instance.exports.archkeep_alloc);
    const pointer = alloc(Math.max(bytes.byteLength, 1));
    new Uint8Array(/** @type {WebAssembly.Memory} */ (instance.exports.memory).buffer).set(
      bytes,
      pointer,
    );
    const evaluate = /** @type {(ptr: number, len: number) => bigint} */ (
      instance.exports.archkeep_evaluate
    );
    const verdict = JSON.parse(readPacked(instance, evaluate(pointer, bytes.byteLength)));
    assert.equal(
      verdict.verdict,
      "unknown",
      `${JSON.stringify(bytes.toString("utf8"))} did not become an unknown verdict`,
    );
    assert.ok(
      typeof verdict.reason === "string" && verdict.reason.length > 0,
      "an unknown verdict names what could not be established",
    );
    assert.deepEqual(verdict.findings, []);
  }
});
